### PR TITLE
Provide recovery_init_sync_method=syncfs.

### PR DESCRIPTION
**Pull request by paul-guo-**
Since commit 2ce439f3 we have opened every file in the data directory
and called fsync() at the start of crash recovery.  This can be very
slow if there are many files, leading to field complaints of systems
taking minutes or even hours to begin crash recovery.

Provide an alternative method, for Linux only, where we call syncfs() on
every possibly different filesystem under the data directory.  This is
equivalent, but avoids faulting in potentially many inodes from
potentially slow storage.

The new mode comes with some caveats, described in the documentation, so
the default value for the new setting is "fsync", preserving the older
behavior.

Reported-by: Michael Brown <michael.brown@discourse.org>
Reviewed-by: Fujii Masao <masao.fujii@oss.nttdata.com>
Reviewed-by: Paul Guo <guopa@vmware.com>
Reviewed-by: Bruce Momjian <bruce@momjian.us>
Reviewed-by: Justin Pryzby <pryzby@telsasoft.com>
Reviewed-by: David Steele <david@pgmasters.net>
Discussion: https://postgr.es/m/11bc2bb7-ecb5-3ad0-b39f-df632734cd81%40discourse.org
Discussion: https://postgr.es/m/CAEET0ZHGnbXmi8yF3ywsDZvb3m9CbdsGZgfTXscQ6agcbzcZAw%40mail.gmail.com

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by paul-guo-** [comment]
I did a simple testing. Here are the steps:

- Populate the tables: 10K aocs tables with 100 columns, create and then insert one tuple.
- fsync=off after this step.
- Finally 2M+ files in the demo cluster. 

Run "sync; echo 3 > /proc/sys/vm/drop_caches" before each test:

gpstop -M immediate and then 
time gpstart -a

Results:

1. Default setting
   real	0m47.969s
2. gpconfig -c recovery_init_sync_method -v syncfs --skipvalidation   
   real	0m5.343s

This test is just a rough testing since in real scenario some file meta info should be in cache (so test 1 should be faster considering this), but my test environment is a VM so the drop_caches operation is not that strictly useful since the vm disk file might be still in the memory of the host OS (so test 1 should be slower considering this). Anyway, from the real testing result and in theory (by code understanding) syncfs is surely faster if there are a lot of pgdata files. The drawback of syncfs is that if multiple pgdata are on the same fs but it seems that  in real scenarios we usully deploy different seg pgdata on different fs, and even multiple pgdata on the same fs, syncfs is usually much faster if the file number is large.

**Comment by paul-guo-** [comment]
Oops. In the above steps, there is a typo. Should be "fsync=on" of course.

File: src/include/utils/unsync_guc_name.h [hunk 1]
```diff
@@ -138,6 +138,7 @@
 		"external_pid_file",
 		"extra_float_digits",
 		"from_collapse_limit",
+		"recovery_init_sync_method",
```
Comment (asimrp) [inline src/include/utils/unsync_guc_name.h:141]: Shouldn't this GUC be consistent across the cluster? What is the reason to add it to the unsync list?

**Review by asimrp — APPROVED** [review]
Thank you for sharing the performance numbers, they look promising.  We should back port this to PR to 6X.

File: src/include/utils/unsync_guc_name.h (hunk 1 above)
Comment (ashwinstar) [reply src/include/utils/unsync_guc_name.h:141]: - GUC `recovery_init_sync_method` should be ideally same across the cluster. But nothing goes wrong if it doesn't from Cluster perspective
- GUC is defined as `PGC_POSTMASTER`, so ideally only way to set is via postgresql.conf file
- sync or unsync list doesn't serve the purpose of which GUCs need to be consistent across the cluster via the file. It can only help in situations where session level setting of GUC needs to be consistent.

So, considering those seems GUC `recovery_init_sync_method` seems fine in unsync list.

**Comment by ashwinstar** [comment]
Considering GPDB use case and number of objects typically always present, I am still trying to understand why we don't wish to have GUC `recovery_init_sync_method` default defined as `syncfs`? If majority of users will have to change the GUC then seems default value is not serving the purpose.
- Please state what's the major downside for `syncfs`?
- In which scenarios `fsync` should be preferred for GPDB users?
- How would users figure out what's best option for them?
- If we feel its safe for single or subset of users to set it to syncfs, then we should be fine for all, right?

**Comment by paul-guo-** [comment]
> Considering GPDB use case and number of objects typically always present, I am still trying to understand why we don't wish to have GUC `recovery_init_sync_method` default defined as `syncfs`? If majority of users will have to change the GUC then seems default value is not serving the purpose.
> 
>     * Please state what's the major downside for `syncfs`?
The document in the original PG patch talks a bit about this.
(The document files are not included in gpdb repo so the change
does not exist in the PR).

```
+        On Linux, <literal>syncfs</literal> may be used instead, to ask the
+        operating system to synchronize the whole file systems that contain the
+        data directory, the WAL files and each tablespace (but not any other
+        file systems that may be reachable through symbolic links).  This may
+        be a lot faster than the <literal>fsync</literal> setting, because it
+        doesn't need to open each file one by one.  On the other hand, it may
+        be slower if a file system is shared by other applications that
+        modify a lot of files, since those files will also be written to disk.
+        Furthermore, on versions of Linux before 5.8, I/O errors encountered
+        while writing data to disk may not be reported to
+        <productname>PostgreSQL</productname>, and relevant error messages may
+        appear only in kernel logs.
```

For the first drawback "it may be slower if a file system is shared by other applications..."
Actually I do not worry much since even we usually deploy multiple segments
on the same host but it seems that it is quite common that different segment pgdata are on
different file systems. Since we do not know how the user deploy the cluster we
are not 100% sure if syncfs is better for them or not.

For the 2nd one, i.e. the Linux kernel requirement of 5.8 - I know some Linux releases
even use low-version kernel but still backports some high-version fixes and even features.
So need to check the os releases if syncfs is correctly supported or not.

I think for the gp6 backport, we need to answer the below questions in the document.
Some things need to do in my mind

- If the OS version does not support that we should additionally disable it when compiling
- Document when 

>     * In which scenarios `fsync` should be preferred for GPDB users?
> 
>     * How would users figure out what's best option for them?
> 
>     * If we feel its safe for single or subset of users to set it to syncfs, then we should be fine for all, right?

**Comment by paul-guo-** [comment]
Re-posting my comment since I just "close with comment" mistakenly.

> Considering GPDB use case and number of objects typically always present, I am still trying to understand why we don't wish to have GUC `recovery_init_sync_method` default defined as `syncfs`? If majority of users will have to change the GUC then seems default value is not serving the purpose.
> 
>     * Please state what's the major downside for `syncfs`?

The document change in the original PG patch talks a bit about this.
(The document files are not included in gpdb repo so the change
does not exist in the PR).

```
+        On Linux, <literal>syncfs</literal> may be used instead, to ask the
+        operating system to synchronize the whole file systems that contain the
+        data directory, the WAL files and each tablespace (but not any other
+        file systems that may be reachable through symbolic links).  This may
+        be a lot faster than the <literal>fsync</literal> setting, because it
+        doesn't need to open each file one by one.  On the other hand, it may
+        be slower if a file system is shared by other applications that
+        modify a lot of files, since those files will also be written to disk.
+        Furthermore, on versions of Linux before 5.8, I/O errors encountered
+        while writing data to disk may not be reported to
+        <productname>PostgreSQL</productname>, and relevant error messages may
+        appear only in kernel logs.
```

For the first drawback "it may be slower if a file system is shared by other applications..."
Actually I do not worry much since even we usually deploy multiple segments
on the same host but it seems that it is quite common that different segment pgdata are on
different file systems. Since we do not know how the user deploy the cluster we
are not 100% sure if syncfs is better for them or not.

For the 2nd one, i.e. the Linux kernel requirement of 5.8 - I know some Linux releases
even use low-version kernel but still backport some high-version fixes and even features.
So need to check the os releases if syncfs is correctly supported or not.

I think for the gp6 version, we need to answer the below questions in the document and/or code
in another PR. Some things need to do in my mind

- If the OS version does not support that we should additionally disable it when compiling
- Document when should select it.

For master branch, we might just want to create a gp7 GA todo work at this moment.

>     * In which scenarios `fsync` should be preferred for GPDB users?
> 
>     * How would users figure out what's best option for them?
> 
>     * If we feel its safe for single or subset of users to set it to syncfs, then we should be fine for all, right?

**Comment by asimrp** [comment]
@paul-guo- if the following strategy is used to deploy multiple primary/mirror segments on a single host, `syncfs` may be slower.  Three mount points - `/data/mirror` to deploy all mirrors and `/data/primary` to deploy all primaries and `/data/fast_ts` as location for user defined tablespaces.  In this case, if only one of several mirrors needs to be synced, `syncfs` may likely be slower, because other segments may be operating on the same mount points and `syncfs` would end up synchronizing a lot more than just that specific mirror's files.  Just thought of mentioning this scenario here.

**Comment by ashwinstar** [comment]
I think its typical to have a primary and mirror (of course of different content) on same filesystem. So, we should definitely test that case out and should have idea about the performance implications. Possibly, given the number of objects present in GPDB, its possible maybe `syncfs` comes still as winner even if its doing more work like syncing primary and mirror data (even if shouldn't).

I dislike pushing the choice to users for this, we should try to do as much as due diligence as possible and choose the default which works for > 80% of GPDB deployments. Other 20% cases, given its GUC can choose something no-default.

**Comment by paul-guo-** [comment]
Probably we could calculate and suggest/set a default guc value in utility code based on the deployment. pgdata volume might not be a factor, if it is large, syncfs wins and if it is small, it is not a big deal to use any method. Maybe we could decide based on how many segments (primary, mirror) on the same fs and the Linux version.

**Comment by paul-guo-** [comment]
> Please can you test the case where you feel `fsync` will be better for GPDB, like many segments on same fs with many AO tables and then report the numbers to see if its really a winner over `syncfs()`?

It's hard to measure since this depends on the disk speed, memory size. With more and more files, `fsync` is slower, especially if there are millions of files, `fsync` has to silly load some files' meta info to memory and then mostly do nothing or probably flush meta info (e.g. timestamp) which is unnecessary for us. `syncfs` quite depends on the total memory size for dirty page cache but as known memory size is usually much less than disk size so mostly `syncfs` should be better in real production environment.

**Comment by ashwinstar** [comment]
> > Please can you test the case where you feel `fsync` will be better for GPDB, like many segments on same fs with many AO tables and then report the numbers to see if its really a winner over `syncfs()`?
> 
> It's hard to measure since this depends on the disk speed, memory size. With more and more files, `fsync` is slower, especially if there are millions of files, `fsync` has to silly load some files' meta info to memory and then mostly do nothing or probably flush meta info (e.g. timestamp) which is unnecessary for us. `syncfs` quite depends on the total memory size for dirty page cache but as known memory size is usually much less than disk size so mostly `syncfs` should be better in real production environment.

I tried to ask the questions in various shapes and forms. Every time answer received is `syncfs()` is better. Then let me ask again, why are we not default setting to `syncfs` instead of `fsync` ?

**Comment by paul-guo-** [comment]
> I tried to ask the questions in various shapes and forms. Every time answer received is `syncfs()` is better. Then let me ask again, why are we not default setting to `syncfs` instead of `fsync` ?

Let me summarize my options here:

- I've done a simple perf testing and syncfs is basically better than fsync in usual cases. And the kernel theory of fsync and syncfs are quite clear to us.

-  If there are many segments on the same fs, syncfs should not be the default. It's not unusual. e.g.  the scenario that there are hundreds of segments on one host so very possible many segments share one fs; or the scenario users use one SAN or network fs for one segment. Also some Linux kernel version does not seem to support syncfs (e.g. rel6). We need to determine a good default value for users.

**Comment by ashwinstar** [comment]
>  If there are many segments on the same fs, syncfs should not be the default. It's not unusual. e.g.  the scenario that there are hundreds of segments on one host so very possible many segments share one fs; or the scenario users use one SAN or network fs for one segment.

I don't think that typical installation setup. Means yes some setups will be that way and those can use GUC to set to fsync as that's preferred for those environments. But majority installs don't seem to fall in that category.

> Also some Linux kernel version does not seem to support syncfs (e.g. rel6). We need to determine a good default value for users.

- So, what happens if we enable `syncfs` as default and its not supported by the platform, does it fallback to fsync?
- We still to think then how to resolve the problem for installs which are on platforms where syncfs is not present, what alternatives do we have?

**Comment by paul-guo-** [comment]
> > If there are many segments on the same fs, syncfs should not be the default. It's not unusual. e.g.  the scenario that there are hundreds of segments on one host so very possible many segments share one fs; or the scenario users use one SAN or network fs for one segment.
> 
> I don't think that typical installation setup. Means yes some setups will be that way and those can use GUC to set to fsync as that's preferred for those environments. But majority installs don't seem to fall in that category.

OK. Seems fine for gp to set syncfs as default but we need to document the details about when we should use fsync instead.

> 
> > Also some Linux kernel version does not seem to support syncfs (e.g. rel6). We need to determine a good default value for users.
> 
>     * So, what happens if we enable `syncfs` as default and its not supported by the platform, does it fallback to fsync?
> 
>     * We still to think then how to resolve the problem for installs which are on platforms where syncfs is not present, what alternatives do we have?

Never mind. I found syncfs support is determined during configuring/compiling so we could set fsync as default in code if the platform does not support that. We should document a bit about this (at least let users know some platforms do not support syncfs).

**Comment by asimrp** [comment]
@paul-guo- let me suggest an experiment (not necessarily you, someone else might do it).  Let's initialize a multi host cluster (say, 1 host for QD, 2 hosts for QEs and 3 primary segments per host) in GCP, with a kernel version that supports `syncfs`.  Initially, attach one [persistent disk](https://cloud.google.com/compute/docs/disks/#pdspecs) to each host.  Thus, all the three primary segments as well as the mirror share the same mount point.  Measure performance of `syncfs`.  And then attach more persistent disk to each host such that there is one dedicated disk for each Greenplum primary segment's data directory.  Mirrors on a host can be deployed on local transient disk.  And then measure `syncfs` performance again.

Would the numbers help us make a decision whether to keep `syncfs` as the default sync method?

**Comment by paul-guo-** [comment]
I took a look at the gpdb installation doc & related utility code and thought again. It seems that we could decide if enable syncfs or not in gpinitstandby. The things we need to consider are:

- performance: if too many segments share the same mount point we use fsync. How to define "too many"? Maybe need to do some simple quick perf testing as Asim said.

- Some OS releases do not support syncfs.

- I found `SyncDataDirectory()` simply skips fsync/syncfs return errors. This seems to be dangerous for `syncfs`. I think if we set `syncfs` we should consider rerunning with `fsync` if the `syncfs` fails.

**Comment by paul-guo-** [comment]
I did further investigation. Here are the findings.

The origin upstream patch says:
"Furthermore, on versions of Linux before 5.8, I/O errors encountered while writing data to disk may not be reported to <productname>PostgreSQL</productname>, and relevant error messages may appear only in kernel logs."

This is true. Presumably the below Linux kernel fix is related to this issue.
```
commit 735e4ae5ba28c886d249ad04d3c8cc097dad6336
Author: Jeff Layton <jlayton@kernel.org>
Date:   Mon Jun 1 21:45:36 2020 -0700

    vfs: track per-sb writeback errors and report them to syncfs

    Patch series "vfs: have syncfs() return error when there are writeback
    errors", v6.
``` 

I checked latest centos7 kernel.spec file and source code and found this fix was not backported. This fix is critical since we want to skip fsync in pg_rewind/pg_basebackup in https://github.com/greenplum-db/gpdb/pull/11881 so it is critical that startup can ensure data is persistent.

I did not see syncfs manpage in centos6. Presumably centos6 do not have syncfs + the fix also. 

I also checked Ubuntu 18.04.5 release note (https://wiki.ubuntu.com/BionicBeaver/ReleaseNotes/ChangeSummary/18.04.5) and did not find this fix also.

In short, main Linux releases for gpdb  does not support reliable syncfs().

Note as I mentioned is that `SyncDataDirectory()` ignores the syncfs() error - this is wrong, but this fixable for us.

In all, I think it is a bit early to move forward the syncfs work. We could revisit this later when the OS releases support syncfs() better.

**Comment by asimrp** [comment]
That means this PR should be merged in with `fsync` as the default recovery sync method.

Does it make sense to probe the kernel version and set recovery sync method to `syncfs` if the kernel version is newer than 5.8?  Such utilities would be `gpinitsystem`, `gprecoverseg`, `gpinitstandby`, etc.

**Comment by paul-guo-** [comment]
> That means this PR should be merged in with `fsync` as the default recovery sync method.

I mean we should close the PR without merging but should create a github issue to avoid forgetting.

> Does it make sense to probe the kernel version and set recovery sync method to `syncfs` if the kernel version is newer than 5.8? Such utilities would be `gpinitsystem`, `gprecoverseg`, `gpinitstandby`, etc.

5.8 is quite new. rhel & ubuntu releases does not support that version to my best of knowledge, but please note Linux OS providers may backport some fixes or features in upstream to their older kernels so judging via kernel version is not a good solution. One solution is to determine for each OS releases when the fix was included in those releases.

**Comment by ashwinstar** [comment]
So, to summarize upstream didn't really solve the problem for back branches and GPDB essentially with `syncfs()`. This problem is very real and super problematic for greenplum deployments. We really need to solve it. So, essentially need to go back to drawing broad and start solving it fresh. Plus, frankly and ideally with recursive fsync or syncfs things were still not in great shape for Greenplum, in terms of performance as majority of files present will be AO/CO related.

Given those I think the approach discussed in thread [1] by Thomas might be more feasible alternative for GPDB (as can only fsync required and only heap files):
```
I also wrote a WAL-and-checkpoint based prototype[2], which, among
other advantages such as being faster, not ignoring errors and not
triggering collateral write-back storms, happens to work on all
operating systems.  On the other hand it requires a somewhat dogmatic
switch in thinking about the meaning of checkpoints (I mean, it
requires humans to promise not to falsify checkpoints by copying
databases around underneath us), which may be hard to sell (I didn't
try very hard), and there may be subtleties I have missed...
```

1] https://www.postgresql.org/message-id/CA+hUKGKHhDNnN6fxf6qrAx9h+mjdNU2Zmx7ztJzFQ0C5=u3QPg@mail.gmail.com

**Comment by paul-guo-** [comment]
@ashwinstar Yes I saw Thomas' s words also. The ideal solution is never fsync/syncfs on pgdata. Frankly speaking I know we need to sync the wal files but for table data part I can not figure out why we do fsync them since we have full page write on.

**Comment by paul-guo-** [comment]
I just found the patch in [1] says
"Typically, this means doing no work at all, because most users run with
full_page_writes=on, meaning that we'll restore the earliest version of
the page and then replay every WAL record on top of it (that is, we'll
never skip a record, and always dirty, write back and eventually sync
the page).

With full_page_writes=off, we get to skip dirtying pages that have
higher LSNs ("BLK_DONE"), but now in that case we'll still queue up a
request to issue an fsync() call in the next checkpoint whenever that
happens."

It seems that for the crash recovery case, if full page writes are enabled really wal sync is enough. Note this is still not enough for us since we have the pg_basebackup & pg_rewind cases. For the pg_basebackup case we need to sync the whole pgdata (this sync is inevitable), for pg_rewind case, currently we do sync the whole pgdata also but we actually could sync those affected files since we ensured clean shutdown before rewinding, see [2] for my PG patch. Then in the startup process, we do sync for the crash recovery case only. We sync the wal files, and do we want to skip sync-ing the data files touched during wal replay? The patch in [1] for the data file sync part is a bit tricky/risky (wal sync part is simple and ok for us). I do not think it is ready for the gp code. Is it possible for us to set full_page_writes as PGC_POSTMASTER and then determine during startup if we want fsync (if full page is off) or sync_wal during startup.

The whole design could be:
  if having backup_label file, skip any sync (the pg_basebackup & pg_rewind cases)
  else if ckpt.full_page_writes is off, sync pgdata
  else sync wal files.

Note pg_basebackup fsync code could be improved also. Currently upstream code does pgdata fsync finally this is not optimal for the case that pgdata has millions of files since it is quite possible some previously written files' meta info (inode, dentry) have been evicted from memory but pg_basebackup has to stupidly reload them to disk again. This could be improved using a separate fsync pthread. This does not seem to be that urgent though - we could launch a discussion on upstream.

[1] https://www.postgresql.org/message-id/CA+hUKGKHhDNnN6fxf6qrAx9h+mjdNU2Zmx7ztJzFQ0C5=u3QPg@mail.gmail.com
[2] https://www.postgresql.org/message-id/flat/7C1703E7-F3F3-43FA-86EB-177C671BF33C@vmware.com

**Comment by paul-guo-** [comment]
Please ignore my previous words about "set full_page_writes as PGC_POSTMASTER". It seems that we could determine via ckpt.full_page_writes event the guc is PGC_SIGHUP.

**Comment by asimrp** [comment]
If data_checksum is enabled, which should be the case in most production systems, full page writes must be enabled.

**Comment by ashwinstar** [comment]
I agree to go with this (with commit message having the scenarios Heikki explained on Slack)
```
The whole design could be:
if having backup_label file, skip any sync (the pg_basebackup & pg_rewind cases)
else if ckpt.full_page_writes is off, sync pgdata
else sync wal files.
```
Ideally I feel lets keep it simple and avoid check for backup_label file as well. I need to check can new WAL not be streamed and written in presence of backup_label file? If can then checking for backup_label file is not right as need to flush the newly arrived wal as well.

**Comment by paul-guo-** [comment]
> If data_checksum is enabled, which should be the case in most production systems, full page writes must be enabled.

Yes full page writes should be on in prod environments but in case users disable that...

**Comment by paul-guo-** [comment]
> The whole design could be:
> if having backup_label file, skip any sync (the pg_basebackup & pg_rewind cases)
> else if ckpt.full_page_writes is off, sync pgdata
> else sync wal files.
> ```
> 
> Ideally I feel lets keep it simple and avoid check for backup_label file as well. I need to check can new WAL not be streamed and written in presence of backup_label file? If can then checking for backup_label file is not right as need to flush the newly arrived wal as well.

backup_label is unlinked soon in startup before xlog redoing. Even there is new wal (by wal receiver. Did not carefully check the code but I suspect no), there is no relay so it should be fine without sync-ing the new wal files additionally (wal receiver does xlog flush).

**Comment by ashwinstar** [comment]
Okay, nice :+1:  Lets go for it, simplify and solve this issue once, for all platforms.
